### PR TITLE
CastSpacesFixer - add space option

### LINKED ISSUE #2853: Extending CastSpacesFixer to remove space after casting
Hi,

what do you thing about extending CastSpacesFixer so it can be configured similarly to ConcatSpaceFixer:

> ``spacing`` (``'none'``, ``'one'``): spacing to apply around concatenation operator;
    defaults to ``'none'``

Both fixers are part of Symfony's codings standars, yet only the later one we can modify.

### PR DESCRIPTION
**Pull request by kubawerlos**
This one resolves #2853.

It's not BC break, original behaviour set by `'cast_spaces' => true` is replaced by default value.

### CODE REVIEW

File: src/Fixer/CastNotation/CastSpacesFixer.php [hunk 1]
```diff
@@ -21,15 +24,36 @@
...
+    private $space;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configure(array $configuration = null)
+    {
+        parent::configure($configuration);
+
+        $this->space = $this->configuration['space'];
```
Comment (keradus) [inline src/Fixer/CastNotation/CastSpacesFixer.php]: no need for aliasing

File: src/Fixer/CastNotation/CastSpacesFixer.php [hunk 2]
```diff
@@ -83,4 +120,20 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
             }
         }
     }
+
+    private function applyFixNoneSpace(Tokens $tokens)
```
Comment (keradus) [inline src/Fixer/CastNotation/CastSpacesFixer.php]: looks almost the same as applyFixSingleSpace

File: src/Fixer/CastNotation/CastSpacesFixer.php [hunk 3]
```diff
@@ -21,15 +24,36 @@
 /**
  * @author Dariusz Rumiński <dariusz.ruminski@gmail.com>
  */
-final class CastSpacesFixer extends AbstractFixer
+final class CastSpacesFixer extends AbstractFixer implements ConfigurationDefinitionFixerInterface
 {
+    const INSIDE_CAST_SPACE_REPLACE_MAP = [
```
Comment (keradus) [inline src/Fixer/CastNotation/CastSpacesFixer.php]: internal

File: src/Fixer/CastNotation/CastSpacesFixer.php (hunk 1 above)
Comment (kubawerlos) [inline src/Fixer/CastNotation/CastSpacesFixer.php]: right, fixed

File: src/Fixer/CastNotation/CastSpacesFixer.php (hunk 2 above)
Comment (kubawerlos) [inline src/Fixer/CastNotation/CastSpacesFixer.php]: ok, I have merged these 2, made it one `if` deeper

File: src/Fixer/CastNotation/CastSpacesFixer.php (hunk 3 above)
Comment (kubawerlos) [inline src/Fixer/CastNotation/CastSpacesFixer.php]: done

File: tests/Fixer/CastNotation/CastSpacesFixerTest.php [hunk 4]
```diff
@@ -21,18 +21,39 @@
...
      * @param string      $expected
      * @param null|string $input
      *
-     * @dataProvider testFixCastsProvider
+     * @dataProvider provideSingleSpaceCases
      */
-    public function testFixCasts($expected, $input = null)
+    public function testFixCastsSingleSpace($expected, $input = null)
     {
+        $this->fixer->configure(['space' => 'single']);
```
Comment (julienfalque) [inline tests/Fixer/CastNotation/CastSpacesFixerTest.php]: Please do not change this test by setting configuration. Add a new test that uses the same data provider instead. This way, tests prove that spaces are added with both default and `single` configuration.

File: src/Fixer/CastNotation/CastSpacesFixer.php [hunk 5]
```diff
@@ -56,31 +71,42 @@ public function isCandidate(Tokens $tokens)
...
-                // force single whitespace after cast token:
-                if ($tokens[$index + 1]->isWhitespace(" \t")) {
-                    // - if next token is whitespaces that contains only spaces and tabs - override next token with single space
-                    $tokens[$index + 1] = new Token([T_WHITESPACE, ' ']);
-                } elseif (!$tokens[$index + 1]->isWhitespace()) {
-                    // - if next token is not whitespaces that contains spaces, tabs and new lines - append single space to current token
-                    $tokens->insertAt($index + 1, new Token([T_WHITESPACE, ' ']));
+                if ($this->configuration['space'] === 'single') {
+                    // force single whitespace after cast token:
+                    if ($tokens[$index + 1]->isWhitespace(" \t")) {
```
Comment (julienfalque) [inline src/Fixer/CastNotation/CastSpacesFixer.php]: There are three `if` nested levels and a `else` now. Can you rewrite those using early `continue`?

File: tests/Fixer/CastNotation/CastSpacesFixerTest.php (hunk 4 above)
Comment (kubawerlos) [inline tests/Fixer/CastNotation/CastSpacesFixerTest.php]: Ok, so there is `testFixCastsWithDefaultConfiguration` and `testFixCastsSingleSpace` now.

File: src/Fixer/CastNotation/CastSpacesFixer.php (hunk 5 above)
Comment (kubawerlos) [inline src/Fixer/CastNotation/CastSpacesFixer.php]: Done

File: src/Fixer/CastNotation/CastSpacesFixer.php [hunk 6]
```diff
@@ -80,7 +90,25 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
                     // - if next token is not whitespaces that contains spaces, tabs and new lines - append single space to current token
                     $tokens->insertAt($index + 1, new Token([T_WHITESPACE, ' ']));
                 }
+            } else {
```
Comment (julienfalque) [inline src/Fixer/CastNotation/CastSpacesFixer.php]: Please use early `continue` instead of `else`.

Comment (kubawerlos) [inline src/Fixer/CastNotation/CastSpacesFixer.php]: Like this? Looks more complicated that way.

File: src/Fixer/CastNotation/CastSpacesFixer.php [hunk 7]
```diff
@@ -56,22 +71,17 @@ public function isCandidate(Tokens $tokens)
...
+            if (!$token->isCast()) {
+                continue;
+            }
+
+            $tokens[$index] = new Token([
+                $token->getId(),
+                strtr($token->getContent(), self::INSIDE_CAST_SPACE_REPLACE_MAP),
+            ]);
 
+            if ($this->configuration['space'] === 'single') {
```
Comment (SpacePossum) [inline src/Fixer/CastNotation/CastSpacesFixer.php]: please make this Yoda as we follow this rule in the repo, after that I'm :+1: for merging this :)
thanks for picking this up

File: src/Fixer/CastNotation/CastSpacesFixer.php [hunk 8]
```diff
@@ -21,15 +24,27 @@
...
+
+    /**
      * {@inheritdoc}
      */
     public function getDefinition()
     {
         return new FixerDefinition(
-            'A single space should be between cast and variable.',
+            'A single space or none should be between cast and variable.',
             [new CodeSample("<?php\n\$bar = ( string )  \$a;\n\$foo = (int)\$b;")]
```
Comment (julienfalque) [inline src/Fixer/CastNotation/CastSpacesFixer.php]: Please add a new code sample with the new option.

Comment (kubawerlos) [inline src/Fixer/CastNotation/CastSpacesFixer.php]: Done

**Comment by keradus** [comment]
Thank you @kubawerlos.

**Comment by SpacePossum** [comment]
thanks @kubawerlos :)

**Pull request closed without merging**